Intermittent failures of test_monotone_frequency

Nightly tests fail occasionally with the following result:

```
======================================================================
FAIL: test_monotone_frequency (test_integration.test_ee_stream_dgemm_mix.TestIntegrationEEStreamDGEMMMix)
Test that agent selects lower frequencies for regions with more
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/test/geopm/test_integration/test_ee_stream_dgemm_mix.py", line 158, in test_monotone_frequency
    msg='Chosen frequency decreased with increasing dgemm fraction')
AssertionError: Chosen frequency decreased with increasing dgemm fraction
```

--- a/src/DGEMMModelRegion.cpp
+++ b/src/DGEMMModelRegion.cpp
@@ -38,7 +38,7 @@
 
 #include "geopm.h"
 #include "Exception.hpp"
-
+#include "Profile.hpp"
 
 #ifdef GEOPM_ENABLE_MKL
 #include <mkl.h>
@@ -97,6 +97,10 @@ namespace geopm
 
     void DGEMMModelRegion::big_o(double big_o_in)
     {
+        geopm::Profile &prof = geopm::Profile::default_profile();
+        uint64_t start_rid = prof.region("geopm_dgemm__model_region_startup", GEOPM_REGION_HINT_IGNORE);
+        prof.enter(start_rid);
+
         if (m_big_o && m_big_o != big_o_in) {
             free(m_matrix_c);
             free(m_matrix_b);
@@ -126,6 +130,7 @@ namespace geopm
             }
         }
         m_big_o = big_o_in;
+        prof.exit(start_rid);
     }
 
     void DGEMMModelRegion::run(void)

--- a/src/ScalingModelRegion.cpp
+++ b/src/ScalingModelRegion.cpp
@@ -42,6 +42,7 @@
 #include "geopm.h"
 #include "geopm_time.h"
 #include "Exception.hpp"
+#include "Profile.hpp"
 
 namespace geopm
 {
@@ -120,6 +121,9 @@ namespace geopm
 
     void ScalingModelRegion::big_o(double big_o_in)
     {
+        geopm::Profile &prof = geopm::Profile::default_profile();
+        uint64_t start_rid = prof.region("geopm_scaling_model_region_startup", GEOPM_REGION_HINT_IGNORE);
+        prof.enter(start_rid);
         m_big_o = big_o_in;
         size_t num_trial = 11;
         size_t median_idx = num_trial / 2;
@@ -138,6 +142,7 @@ namespace geopm
         m_num_atom = big_o_in / median_atom_time;
         m_num_atom = m_num_atom ? m_num_atom : 1;
         m_norm = 1.0 / m_num_atom;
+        prof.exit(start_rid);
     }
 
     void ScalingModelRegion::run(void)

--- a/src/StreamModelRegion.cpp
+++ b/src/StreamModelRegion.cpp
@@ -37,6 +37,7 @@
 
 #include "geopm.h"
 #include "Exception.hpp"
+#include "Profile.hpp"
 
 namespace geopm
 {
@@ -71,6 +72,10 @@ namespace geopm
 
     void StreamModelRegion::big_o(double big_o_in)
     {
+        geopm::Profile &prof = geopm::Profile::default_profile();
+        uint64_t start_rid = prof.region("geopm_stream_model_region_startup", GEOPM_REGION_HINT_IGNORE);
+        prof.enter(start_rid);
+
         if (m_big_o && m_big_o != big_o_in) {
             free(m_array_c);
             free(m_array_b);
@@ -100,6 +105,8 @@ namespace geopm
             }
         }
         m_big_o = big_o_in;
+
+        prof.exit(start_rid);
     }
 
     void StreamModelRegion::run(void)

--- a/test_integration/Makefile.mk
+++ b/test_integration/Makefile.mk
@@ -57,7 +57,7 @@ test_integration_test_profile_policy_CXXFLAGS = $(AM_CXXFLAGS) -std=c++11
 endif
 
 include test_integration/test_ee_short_region_slop.mk
-include test_integration/test_ee_stream_dgemm_mix.mk
+include test_integration/test_ee_timed_scaling_mix.mk
 include test_integration/test_enforce_policy.mk
 include test_integration/test_omp_outer_loop.mk
 include test_integration/test_scaling_region.mk

--- a/test_integration/__init__.py
+++ b/test_integration/__init__.py
@@ -39,7 +39,7 @@ sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
 from test_integration import geopm_context
 from test_integration.geopm_test_integration import *
 from test_integration.test_omp_outer_loop import *
-from test_integration.test_ee_stream_dgemm_mix import *
+from test_integration.test_ee_timed_scaling_mix import *
 from test_integration.test_enforce_policy import *
 from test_integration.test_profile_policy import *
 from test_integration.test_plugin_static_policy import *

--- a/test_integration/test_ee_timed_scaling_mix.cpp
+++ b/test_integration/test_ee_timed_scaling_mix.cpp
@@ -59,54 +59,53 @@ int main(int argc, char **argv)
             }
         }
     }
-    std::unique_ptr<geopm::ModelRegion> spin_model(geopm::ModelRegion::model_region("spin", 0.075, is_verbose));
-    std::unique_ptr<geopm::ModelRegion> short_model(geopm::ModelRegion::model_region("spin", 0.01, is_verbose));
-    uint64_t short_region_id = 0;
-    err = geopm_prof_region("short_region", GEOPM_REGION_HINT_UNKNOWN, &short_region_id);
-    if (err) {
-        throw geopm::Exception("test_ee_stream_dgemm_mix", err, __FILE__, __LINE__);
-    }
 
-    int repeat = 300;
-    double dgemm_factor = 17;
-    double stream_factor = 1.0;
+    std::string test_name = "test_ee_timed_scaling_mix";
+    std::unique_ptr<geopm::ModelRegion> ignore_model(geopm::ModelRegion::model_region("spin", 0.075, is_verbose));
+    uint64_t ignore_region_id = 0;
+    err = geopm_prof_region("ignore", GEOPM_REGION_HINT_IGNORE, &ignore_region_id);
+    if (err) {
+        throw geopm::Exception(test_name, err, __FILE__, __LINE__);
+    }
+    int repeat = 100;
+    double scaling_factor = 1;
+    double timed_factor = 1;
     int num_mix = 5;
     double mix_factor = 1.0 / (num_mix - 1);
     for (int mix_idx = 0; mix_idx != num_mix; ++mix_idx) {
-        int stream_idx = num_mix - 1 - mix_idx;
-        int dgemm_idx = mix_idx;
-        double stream_big_o = stream_factor * mix_factor * stream_idx;
-        double dgemm_big_o = dgemm_factor * mix_factor * dgemm_idx;
-        std::unique_ptr<geopm::ModelRegion> stream_model(geopm::ModelRegion::model_region("stream-unmarked", stream_big_o, is_verbose));
-        std::unique_ptr<geopm::ModelRegion> dgemm_model(geopm::ModelRegion::model_region("dgemm-unmarked", dgemm_big_o, is_verbose));
+        int timed_idx = num_mix - 1 - mix_idx;
+        int scaling_idx = mix_idx;
+        double timed_big_o = timed_factor * mix_factor * timed_idx;
+        double scaling_big_o = scaling_factor * mix_factor * scaling_idx;
+        std::unique_ptr<geopm::ModelRegion> timed_model(geopm::ModelRegion::model_region("timed_scaling-unmarked", timed_big_o, is_verbose));
+        std::unique_ptr<geopm::ModelRegion> scaling_model(geopm::ModelRegion::model_region("scaling-unmarked", scaling_big_o, is_verbose));
         char region_name[NAME_MAX];
         region_name[NAME_MAX - 1] = '\0';
-        snprintf(region_name, NAME_MAX - 1, "stream-%.2f-dgemm-%.2f", stream_big_o, dgemm_big_o);
+        snprintf(region_name, NAME_MAX - 1, "timed-%.2f-scaling-%.2f", timed_big_o, scaling_big_o);
         uint64_t region_id = 0;
         err = geopm_prof_region(region_name, GEOPM_REGION_HINT_UNKNOWN, &region_id);
         if (err) {
-            throw geopm::Exception("test_ee_stream_dgemm_mix", err, __FILE__, __LINE__);
+            throw geopm::Exception(test_name, err, __FILE__, __LINE__);
         }
         for (int rep_idx = 0; rep_idx != repeat; ++rep_idx) {
             err = geopm_prof_enter(region_id);
             if (err) {
-                throw geopm::Exception("test_ee_stream_dgemm_mix", err, __FILE__, __LINE__);
+                throw geopm::Exception(test_name, err, __FILE__, __LINE__);
             }
-            stream_model->run();
-            dgemm_model->run();
+            timed_model->run();
+            scaling_model->run();
             err = geopm_prof_exit(region_id);
             if (err) {
-                throw geopm::Exception("test_ee_stream_dgemm_mix", err, __FILE__, __LINE__);
+                throw geopm::Exception(test_name, err, __FILE__, __LINE__);
             }
-            spin_model->run();
-            err = geopm_prof_enter(short_region_id);
+            err = geopm_prof_enter(ignore_region_id);
             if (err) {
-                throw geopm::Exception("test_ee_stream_dgemm_mix", err, __FILE__, __LINE__);
+                throw geopm::Exception(test_name, err, __FILE__, __LINE__);
             }
-            short_model->run();
-            err = geopm_prof_exit(short_region_id);
+            ignore_model->run();
+            err = geopm_prof_exit(ignore_region_id);
             if (err) {
-                throw geopm::Exception("test_ee_stream_dgemm_mix", err, __FILE__, __LINE__);
+                throw geopm::Exception(test_name, err, __FILE__, __LINE__);
             }
             MPI_Barrier(MPI_COMM_WORLD);
         }

--- a/test_integration/test_ee_timed_scaling_mix.mk
+++ b/test_integration/test_ee_timed_scaling_mix.mk
@@ -29,15 +29,15 @@
 #  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #
 
-EXTRA_DIST += test_integration/test_ee_stream_dgemm_mix.py
+EXTRA_DIST += test_integration/test_ee_timed_scaling_mix.py
 
 if ENABLE_MPI
-noinst_PROGRAMS += test_integration/test_ee_stream_dgemm_mix
-test_integration_test_ee_stream_dgemm_mix_SOURCES = test_integration/test_ee_stream_dgemm_mix.cpp
-test_integration_test_ee_stream_dgemm_mix_SOURCES += $(model_source_files)
-test_integration_test_ee_stream_dgemm_mix_LDADD = libgeopm.la $(MATH_LIB) $(MPI_CLIBS)
-test_integration_test_ee_stream_dgemm_mix_LDFLAGS = $(AM_LDFLAGS) $(MPI_CLDFLAGS) $(MATH_CLDFLAGS)
-test_integration_test_ee_stream_dgemm_mix_CXXFLAGS = $(AM_CXXFLAGS) $(MPI_CFLAGS) -D_GNU_SOURCE -std=c++11 $(MATH_CFLAGS)
+noinst_PROGRAMS += test_integration/test_ee_timed_scaling_mix
+test_integration_test_ee_timed_scaling_mix_SOURCES = test_integration/test_ee_timed_scaling_mix.cpp
+test_integration_test_ee_timed_scaling_mix_SOURCES += $(model_source_files)
+test_integration_test_ee_timed_scaling_mix_LDADD = libgeopm.la $(MATH_LIB) $(MPI_CLIBS)
+test_integration_test_ee_timed_scaling_mix_LDFLAGS = $(AM_LDFLAGS) $(MPI_CLDFLAGS) $(MATH_CLDFLAGS)
+test_integration_test_ee_timed_scaling_mix_CXXFLAGS = $(AM_CXXFLAGS) $(MPI_CFLAGS) -D_GNU_SOURCE -std=c++11 $(MATH_CFLAGS)
 else
-EXTRA_DIST += test_integration/test_ee_stream_dgemm_mix.cpp
+EXTRA_DIST += test_integration/test_ee_timed_scaling_mix.cpp
 endif

--- a/test_integration/test_ee_timed_scaling_mix.py
+++ b/test_integration/test_ee_timed_scaling_mix.py
@@ -31,7 +31,8 @@
 #  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #
 
-"""Test the energy efficient agent by doing a frequency stream/dgemm mix.
+"""Test the energy efficient agent by running a mix of frequency-
+   sensitive and frequency-insensitive regions.
 
 """
 
@@ -69,7 +70,7 @@ class AppConf(object):
 
         """
         script_dir = os.path.dirname(os.path.realpath(__file__))
-        return os.path.join(script_dir, '.libs', 'test_ee_stream_dgemm_mix')
+        return os.path.join(script_dir, '.libs', 'test_ee_timed_scaling_mix')
 
     def get_exec_args(self):
         return []
@@ -78,7 +79,7 @@ class AppConf(object):
 @util.skip_unless_cpufreq()
 @util.skip_unless_optimized()
 @util.skip_unless_run_long_tests()
-class TestIntegrationEEStreamDGEMMMix(unittest.TestCase):
+class TestIntegrationEETimedScalingMix(unittest.TestCase):
     @classmethod
     def setUpClass(cls):
         """Create launcher, execute benchmark and set up class variables.
@@ -86,14 +87,14 @@ class TestIntegrationEEStreamDGEMMMix(unittest.TestCase):
         """
         sys.stdout.write('(' + os.path.basename(__file__).split('.')[0] +
                          '.' + cls.__name__ + ') ...')
-        test_name = 'test_ee_stream_dgemm_mix'
+        test_name = 'test_ee_timed_scaling_mix'
         cls._report_path = test_name + '.report'
         cls._trace_path = test_name + '.trace'
         cls._skip_launch = _g_skip_launch
         cls._keep_files = os.getenv('GEOPM_KEEP_FILES') is not None
         cls._agent_conf_path = test_name + '-agent-config.json'
         geopmpy.error.exc_clear()
-        if  not cls._skip_launch:
+        if not cls._skip_launch:
             num_node = 2
             num_rank = 2
             min_freq = geopm_test_launcher.geopmread("CPUINFO::FREQ_MIN board 0")
@@ -131,7 +132,7 @@ class TestIntegrationEEStreamDGEMMMix(unittest.TestCase):
         """
         report = geopmpy.io.RawReport(self._report_path)
         host_names = report.host_names()
-        mix_name_re = re.compile(r'stream-[0-9]*\.[0-9]*-dgemm-[0-9]*\.[0-9]*')
+        mix_name_re = re.compile(r'timed-[0-9]*\.[0-9]*-scaling-[0-9]*\.[0-9]*')
         num_re = re.compile(r'[0-9]*\.[0-9]*')
         for host_name in report.host_names():
             result = []
@@ -168,6 +169,19 @@ class TestIntegrationEEStreamDGEMMMix(unittest.TestCase):
                     self.assertTrue('requested-online-frequency' not in region)
                     found_barrier = True
             self.assertTrue(found_barrier)
+
+    def test_skip_ignore_regions(self):
+        """Test that agent does not learn for regions declared with the ignore hint.
+
+        """
+        report = geopmpy.io.RawReport(self._report_path)
+        host_names = report.host_names()
+        for host_name in report.host_names():
+            for region_name in report.region_names(host_name):
+                if region_name in ['ignore']:
+                    region = report.raw_region(host_name, region_name)
+                    self.assertTrue('requested-online-frequency' not in region)
+                    found_barrier = True
 
 
 if __name__ == '__main__':

